### ISSUE TITLE
Memory use issues

### ISSUE THREAD

**Issue by jliljebl** [jliljebl/flowblade#327]
This issues has been created as merged issue for multiple similar reports:
#300 #235

**Comment by jliljebl** [comment]
issue #300

OS: 4.4.0-57-generic GNU/Linux

Flowblade starts fast...
I have been rendering between markers (application seemed more responsive/faster to do that rather than moving parts of the movie around)

but I noticed it got steadily slower with each render

I checked $top
and 3 instances of python were using a total of 50% of RAM and memory on my machine
workaround(?):
shutdown flowblade between renders/edits.

desired:
mem is kept clean/empty after a render or edit to keep flowblade speedy

**Comment by jliljebl** [comment]
Issue #235

Seem that there are reports of high memory usage when rendering long files.

https://sourceforge.net/p/mlt/bugs/242/
https://plus.google.com/u/0/+PeterZemlyanukha/posts/QUzBDbj1nbK
@out-of-order
out-of-order commented 11 days ago

I unfortunately also have problems with high RAM usage.
I'm using debian stretch and MLT 1:6.4.1-dmo1 from the multimaedia repos.
I have 6 GB RAM and a 4GB swap file.
The desktop takes about 0,25GB
After 2 minutes of rendered film material the RAM is full and tha swap file to the half so the system is not responsive anymore.
ffmpeg is 10:3.2.2-dmo3
@out-of-order
out-of-order commented 11 days ago

setting the same fps for render doesn't raise the memory in my case.

**Comment by dvdlvr** [comment]
It looks like the file descriptors of files previously added to the project remain open even when the file has been removed from the project.
If you add and remove the same file from the project multiple times all the file pointers remain open and the memory use increases every time you add a file.
VmSize:	 2167020 kB
VmSize:	 2284912 kB
VmSize:	 2434448 kB
VmSize:	 2647964 kB
VmSize:	 2795972 kB

I will investigate this further.

**Comment by jliljebl** [comment]
The first thing to look at is that I'm holding references to some objects by design in module "mltrefhold.py" because at least earlier all MLT crashes were related to releasing objects by MLTs homecooked refcount system.

If you comment out all these, does the memory issue go away?

36:import mltrefhold
224:        mltrefhold.hold_ref(self.mlt_filter)
300:            mltrefhold.hold_ref(mlt_filter)

Flowblade/projectdata.py
40:import mltrefhold

Flowblade/sequence.py
34:import mltrefhold
135:        mltrefhold.hold_ref(gain_filter)
144:        mltrefhold.hold_ref(self.vectorscope) # ?? is this just some anti-crash hack attempt that was not removed
148:        mltrefhold.hold_ref(self.rgbparade) # ?? is this just some anti-crash hack attempt that was not removed
274:        mltrefhold.hold_ref(transition)
283:        mltrefhold.hold_ref(gain_filter)
339:        mltrefhold.hold_ref(pan_filter)
385:        mltrefhold.hold_ref(producer)
410:        mltrefhold.hold_ref(producer)
797:        mltrefhold.hold_ref(watermark)

Flowblade/proxyediting.py
39:import mltrefhold
110:            mltrefhold.hold_ref(file_producer) # this may or may not be needed to avoid crashes

Flowblade/keyevents.py
44:import mltrefhold

Flowblade/audiomonitoring.py
46:import mltrefhold
284:    mltrefhold.hold_ref(audio_level_filter)

Flowblade/render.py
48:import mltrefhold
389:        mltrefhold.hold_ref(motion_producer)

Flowblade/mlttransitions.py
31:import mltrefhold
210:        mltrefhold.hold_ref(transition)
482:    mltrefhold.hold_ref(transition)

Flowblade/patternproducer.py
39:import mltrefhold
157:    mltrefhold.hold_ref(producer)
164:    mltrefhold.hold_ref(producer)
169:    mltrefhold.hold_ref(producer)
229:        mltrefhold.hold_ref(producer)
247:        mltrefhold.hold_ref(producer)
260:        mltrefhold.hold_ref(producer)
281:        mltrefhold.hold_ref(producer)
308:        mltrefhold.hold_ref(producer)

**Comment by dvdlvr** [comment]
The hold_ref doesn't seem to be the problem, but it's been hard to follow how the code flows.
The code is very well organized, but it does mean that you have to look in dozens of files to find out what is involved in a simple action.
No worries, I'll get there, but it will take some time before I know the code and mlt and all the datastructures well enough to be more efficient.

**Comment by jliljebl** [comment]
Your view that hold_ref is not the culprit  is useful informations already, and this is likely to require longer investigation, so no hurry. 

If we don't make it for next release its okay since there is a workable workaround for this issue: save your work, close app, open app, load work. Takes only 20sec and is effective, so this isn't such a deal braker even though it is a minor embarrassment to not know the cause.

The code base is attempting to be as modular as possible, and yeah, that can make it a bit hard to follow. However I don't think some overarching architectural pattern could be forced on here, in fact having minimal amount of "structure" and maximum amount of module decoupling is very likely to create a more maintainable code base; features can be added and bugs fixed with less risk of creating new bugs as a side effect. 

But if there is less structure, there is also less patterns and less "understanding" that can be achieved to help navigating the code base.

I will now add two docs in repo next week that can help to get an overview of the code base: a list of modules grouped by function and a shot description of some central ideas present in code base, this last document will be expanded gradually.

**Comment by dvdlvr** [comment]
@janne: I haven't found yet where exactly the problem is (Or how to solve it) but here are some significant intermediate findings. Maybe this gives you an idea.

The function resync.clip_added_to_timeline is executed multiple times.
It's executed when you change the default profile to the file profile:
clip_added_to_timeline called by append_clip
At this point there are no named file descriptors open for the project.

It's executed when you open a clip in the clip monitor (Or probably more exactly when you open the clip monitor):
clip_added_to_timeline called by _insert_clip (Even though we only opened the Clip Monitor and have not yet added anything to the timeline)
It's also executed when you add the clip to the timeline.
Every time the clip_added_to_timeline is called by _insert_clip it adds 2 open file descriptors to the project and increases the amount of memory used.
Just switching between the project monitor (nothing happens) and then the clip monitor will add more open file descriptors and increase memory used - even if nothing is done to the project.

You can inspect this as follows:
1) Get the pid of the running program
2) grep VmSize /proc/[PID]/status # For memory used
3) ls -l /proc/[PID]/fd # Will list all open files. Check for the video file name in there once placed in clip monitor or timeline

There seems to be a maximum of 24 fd's open. Once this is reached, going back and forth between project and clip monitor doesn't open any more. Not sure if this is a python, OS or Flowblade restriction.

Removing the clips you added to the timeline does not close off the file descriptors that where opened when adding the clips and does not clear memory.

When you drag a clip directly from the bin to the timeline it also calls _insert_clip and opens 2 fd's every time you do that.

I haven't found yet why it always opens a multiple of 2 file descriptors at a time. Maybe one for video and one for audio?

**Comment by jliljebl** [comment]
@dvdlvr Thanks for keeping up with this.

I'm currently tied up in making the AppImage work, so I can't look at this carefully atm.

Also this is becoming clearly 1.14 material, I'm putting out 1.12 in few weeks and any fixes for this are clearly going to destabilizing. 

Before 1.12 I'm going to have to:
- create working Appimage for 1.10
- decide on exact names for 2 new tools
- do some more patches for 2 new tools
- test the new tools throughly
- update docs for tools
- test other new stuff at least once
- inform Mario on text freeze and update translations when i get them back.
- create .deb and Appimage for 1.12

This is still going to take maybe 2 weeks, might even take a bit longer.

After 1.12 is out I'll try to understand the Issues here. The 1.14 is probably coming out as late as August so this can fixed and tested for that with plenty of time available.

**Comment by dvdlvr** [comment]
@janne No worries. I'll keep on working on this in the meantime, unless you want help with something else first.

**Comment by jliljebl** [comment]
@dvdlvr No, theres mostly getting the AppImage and new tools finished, so there isn't much else you can do for 1.12.

**Comment by Windsaw** [comment]
I wanted to ask if this issue is still being worked on.
I am using Flowblade 1.16 on a Debian testing system.
To verify I downloaded the current git and tried it with that.

I have several segments about 12 minutes in length with many cuts and as many as 12 different original video files.
Rendering starts and uses up all of my 16 GB of ram and then swaps. Around that time the computer slows down to a halt. If I limit the amount of memory Flowblade can use then rendering eventually crashes.
It does not matter what resolution or codec I use.
The rendering progress bar never goes beyond 0%.
I went back to an old (similar) project that I successfully rendered in Nov 2017. It seems to have the same problem now. (I stopped after it was at 12GB memory usage)
Apart from some Debian updates and a newer harddrive the system is the same as it was back then.

**Comment by out-of-order** [comment]
I had RAM problems when the source fps wasn't identical than destination fps.
Rendering 50 fps from the cam to 25 fps resulted in crazy RAM usage.
Set the renderer to equal fps.
If you want to reduce the fps, convert the soruce material before import or as a plugin filter in flowblade (if available).

**Comment by jliljebl** [comment]
This is *active* but not currently actively being worked as it is almost certainly an MLT issue, and currently all work is done purely on Flowblade side of things, It has been medium term plan to move more focus on MLT and start doing patches there again after a long hiatus, but I haven't got around to doing that yet.

@dvdlvr is right that it would be useful to know if issue is *Flowblade* specific or does rendering with *melt* or *KDEnlive* produce similar problems, so if we can get info from affected systems on that, we would be one small step closer in getting results here.

**Comment by Windsaw** [comment]
Thank you for your suggestion!
I installed melt (same version as shown for MLT) and followed your instructions.
Rendering from command line works. Memory usage is constant and from glancing over it the finished video seems to be as intended.
I am going to use this as a workaround for now.

For detailed info:
I am creating a tutorial video. Raw material is 32GB of videos in 48 files although I don't use them all. (since AFAIK Flowblade does not have time lapse functionality I prerendered all files with various speeds to cut them into the video) Source material are a mixture of Canon EOS video recordings, from my Android smartphone and screencasts.
So far I only made simple cuts, no transitions or filters.
Although it works on command line I have attached the XML file of one of the segments.

[Sunstone_Slipcase_02.zip](https://github.com/jliljebl/flowblade/files/2412959/Sunstone_Slipcase_02.zip)
